apollo-federation/src/operation/mod.rs

```diff
@@ -1443,33 +906,46 @@ mod fragment_spread_selection {
         pub(crate) selection_id: SelectionId,
     }
 
-    impl HasSelectionKey for FragmentSpreadData {
-        fn key(&self) -> SelectionKey {
+    impl PartialEq for FragmentSpread {
+        fn eq(&self, other: &Self) -> bool {
+            self.key() == other.key()
+        }
+    }
+
+    impl Eq for FragmentSpread {}
+
+    impl FragmentSpread {
+        pub(super) fn directives_mut(&mut self) -> &mut DirectiveList {
+            &mut self.directives
+        }
+    }
+
+    impl HasSelectionKey for FragmentSpread {
+        fn key(&self) -> SelectionKey<'_> {
             if is_deferred_selection(&self.directives) {
```

Comment:
This is a tradeoff right now: we call `key()` much more frequently, and for inline fragments and fragment spreads it has to search through the directive list. In many cases it'll be free (because the directive list is `None`), but in some cases it will have to walk the list and do a few string comparisons. It could potentially be optimized by storing a `.has_defer` boolean in `DirectiveList`, but I didn't see it on flame graphs yet, so that could be overengineering it.